src/main/java/frc/robot/Constants.java

```diff
@@ -104,6 +104,99 @@ public static final class ModuleConstants {
+        public static final double EXTRA_SAFETY_DISTANCE = 0.1;
+
+        public enum Chain {
+            CHAIN_0(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_120(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_240(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_60(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_180(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_300(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0));
+
+            private final Translation2d negEdge;
+            private final Translation2d posEdge;
+            private final Translation2d centerStage;
+
+            Chain(Translation2d negEdge, Translation2d posEdge, Translation2d centerStage) {
+                this.negEdge = negEdge;
+                this.posEdge = posEdge;
+                this.centerStage = centerStage;
+            }
+
+            public Translation2d getNegEdge() {
+                return negEdge;
+            }
+
+            public Translation2d getPosEdge() {
+                return posEdge;
+            }
+
+            private boolean inRange(Translation2d robotTranslation) {
+                double robotStageAngle =
+                        Math.atan(robotTranslation.minus(centerStage).getX() / robotTranslation.minus(centerStage).getY());
+                double negAngle = Math.atan(negEdge.minus(centerStage).getX() / negEdge.minus(centerStage).getY());
+                double posAngle = Math.atan(posEdge.minus(centerStage).getX() / posEdge.minus(centerStage).getY());
+
+                if (posAngle > negAngle) return robotStageAngle < posAngle && robotStageAngle > negAngle;
+                return (robotStageAngle > posAngle && robotStageAngle > negAngle)
+                        || (robotStageAngle < posAngle && robotStageAngle < negAngle);
+            }
+
+            public static Chain getBestChain(Translation2d robotTranslation) {
+                Chain bestChain;
+                //recognise which chain we want to climb on
+                if (isBlueAlliance()) {
+                    if (Chain.CHAIN_0.inRange(robotTranslation)) bestChain = Chain.CHAIN_0;
+                    else if (Chain.CHAIN_120.inRange(robotTranslation)) bestChain = Chain.CHAIN_120;
+                    else bestChain = Chain.CHAIN_240;
+                } else {
+                    if (Chain.CHAIN_60.inRange(robotTranslation)) bestChain = Chain.CHAIN_60;
+                    else if (Chain.CHAIN_180.inRange(robotTranslation)) bestChain = Chain.CHAIN_180;
+                    else bestChain = Chain.CHAIN_300;
+                }
+                return bestChain;
+            }
+
+            public static Translation2d getProjection(Translation2d robotPose, Chain chain) {
```

Comment:
What does this function do? Documentation documentation documentation.
Also, I'm not entirely sure that a `T.f(..., t)` is better than `t.f(...)` -- it very much depends on semantics that are very unclear here.